components/director/pkg/operation/directive.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 The Compass Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package operation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+
+	gqlgen "github.com/99designs/gqlgen/graphql"
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+	"github.com/kyma-incubator/compass/components/director/pkg/persistence"
+)
+
+const ModeParam = "mode"
+
+type directive struct {
+	transact  persistence.Transactioner
+	scheduler Scheduler
+}
+
+// NewDirective creates a new handler struct responsible for the Async directive business logic
+func NewDirective(transact persistence.Transactioner, scheduler Scheduler) *directive {
+	return &directive{
+		transact:  transact,
+		scheduler: scheduler,
+	}
+}
+
+// HandleOperation enriches the request with an Operation information when the requesting mutation is annotated with the Async directive
+func (d *directive) HandleOperation(ctx context.Context, _ interface{}, next gqlgen.Resolver, op graphql.OperationType) (res interface{}, err error) {
+	resCtx := gqlgen.GetResolverContext(ctx)
+	mode, ok := resCtx.Args[ModeParam].(*graphql.OperationMode)
+	if !ok {
+		return nil, apperrors.NewInternalError(fmt.Sprintf("could not get %s parameter", ModeParam))
+	}
+
+	ctx = SaveModeToContext(ctx, *mode)
```

Comment:
We rely on knowing what mode the current operation is in two places in the repo layer (we do mainly checks of the sort `if operation.ModeFromCtx(ctx) == graphql.OperationModeAsync {`).

If we remove the mode from the context we could  achieve the same check logic by looking for an operation object in the context, but this breaks when there are multiple requests (one is async and another is sync). We need to know for every single request what operation mode we're in and we can't infer that from the operations in the context.

In this sense we need to keep the operation mode.

Reply:
i didnt understand why it breaks with sync requests/still dont see the purpose of mode. There is no operation on the context for them at all == sync. 


If you take a step back though, Ideally since you dont have related resouces with this implementation it does not make sense to pass the op in the context at all. You can just set the ready field here (keep the timestamps in the storage)